File: apps/backend/app/services/assignment_service.py
from datetime import datetime, timezone
from typing import Any

from app.core.exceptions import NotFoundError, ValidationError
from app.core.logging import get_logger
from app.repositories.assignment_repository import AssignmentRepository

logger = get_logger(__name__)


class AssignmentService:
    VALID_STATUSES = {"pending", "in_progress", "completed", "cancelled", "overdue"}
    VALID_PRIORITIES = {"low", "medium", "high", "urgent"}

    def __init__(self) -> None:
        self.repository = AssignmentRepository()

    async def _auto_mark_overdue(self, user_id: str) -> None:
        try:
            affected = await self.repository.mark_overdue_assignments(user_id)
            if affected > 0:
                logger.info("Marked %d assignments as overdue for user %s", affected, user_id)
        except Exception as e:
            logger.warning("Failed to auto-mark overdue assignments: %s", e)

    async def list_assignments(
        self, user_id: str, params: dict[str, Any]
    ) -> dict[str, Any]:
        await self._auto_mark_overdue(user_id)
        page = params.pop("page", 1)
        page_size = params.pop("page_size", 20)
        sort_by = params.pop("sort_by", "created_at")
        sort_order = params.pop("sort_order", "desc")

        rows, total = await self.repository.list_assignments(
            user_id, page, page_size, sort_by, sort_order, params
        )
        return {
            "data": rows,
            "total": total,
            "page": page,
            "page_size": page_size,
            "total_pages": max(1, (total + page_size - 1) // page_size),
        }

    async def get_assignment(self, assignment_id: str, user_id: str) -> dict[str, Any]:
        await self._auto_mark_overdue(user_id)
        assignment = await self.repository.get_assignment(assignment_id, user_id)
        if not assignment:
            raise NotFoundError("Assignment not found")
        return assignment

    async def create_assignment(self, user_id: str, data: dict[str, Any]) -> dict[str, Any]:
        self._validate(data)
        return await self.repository.create_assignment(user_id, data)

    async def update_assignment(
        self, assignment_id: str, user_id: str, data: dict[str, Any]
    ) -> dict[str, Any]:
        existing = await self.repository.get_assignment(assignment_id, user_id)
        if not existing:
            raise NotFoundError("Assignment not found")
        clean = {k: v for k, v in data.items() if v is not None}
        if clean:
            self._validate(clean, partial=True)
        return await self.repository.update_assignment(assignment_id, user_id, clean)

    async def delete_assignment(self, assignment_id: str, user_id: str) -> None:
        existing = await self.repository.get_assignment(assignment_id, user_id)
        if not existing:
            raise NotFoundError("Assignment not found")
        await self.repository.delete_assignment(assignment_id, user_id)

    async def duplicate_assignment(self, assignment_id: str, user_id: str) -> dict[str, Any]:
        existing = await self.repository.get_assignment(assignment_id, user_id)
        if not existing:
            raise NotFoundError("Assignment not found")
        payload = {
            "title": f"{existing['title']} (Copy)",
            "description": existing.get("description"),
            "subject_id": existing.get("subject_id"),
            "priority": existing.get("priority", "medium"),
            "status": "pending",
            "due_date": existing.get("due_date"),
            "estimated_study_time": existing.get("estimated_study_time"),
            "tags": existing.get("tags"),
            "notes": existing.get("notes"),
            "is_pinned": False,
            "is_favorite": False,
        }
        return await self.repository.create_assignment(user_id, payload)

    async def bulk_action(
        self, ids: list[str], user_id: str, action: str, value: Any = None
    ) -> list[dict[str, Any]]:
        if action == "delete":
            await self.repository.bulk_delete(ids, user_id)
            return []
        elif action == "complete":
            return await self.repository.bulk_update(ids, user_id, {
                "status": "completed",
                "completion_date": datetime.now(timezone.utc).isoformat(),
            })
        elif action == "archive":
            return await self.repository.bulk_update(ids, user_id, {"is_archived": True})
        elif action == "restore":
            return await self.repository.bulk_update(ids, user_id, {"is_archived": False})
        elif action == "pin":
            return await self.repository.bulk_update(ids, user_id, {"is_pinned": True})
        elif action == "unpin":
            return await self.repository.bulk_update(ids, user_id, {"is_pinned": False})
        elif action == "favorite":
            return await self.repository.bulk_update(ids, user_id, {"is_favorite": True})
        elif action == "unfavorite":
            return await self.repository.bulk_update(ids, user_id, {"is_favorite": False})
        elif action == "set_status" and value:
            return await self.repository.bulk_update(ids, user_id, {
                "status": str(value),
                **({"completion_date": datetime.now(timezone.utc).isoformat()} if str(value) == "completed" else {}),
            })
        elif action == "set_priority" and value:
            return await self.repository.bulk_update(ids, user_id, {"priority": str(value)})
        else:
            raise ValidationError(f"Unknown bulk action: {action}")

    async def get_attachments(self, assignment_id: str, user_id: str) -> list[dict[str, Any]]:
        assignment = await self.repository.get_assignment(assignment_id, user_id)
        if not assignment:
            raise NotFoundError("Assignment not found")
        return await self.repository.get_attachments(assignment_id)

    async def get_stats(self, user_id: str) -> dict[str, Any]:
        await self._auto_mark_overdue(user_id)
        return await self.repository.get_stats(user_id)

    def _validate(self, data: dict[str, Any], partial: bool = False) -> None:
        if not partial or "status" in data:
            status = data.get("status")
            if status and status not in self.VALID_STATUSES:
                raise ValidationError(f"Invalid status: {status}")
        if not partial or "priority" in data:
            priority = data.get("priority")
            if priority and priority not in self.VALID_PRIORITIES:
                raise ValidationError(f"Invalid priority: {priority}")

File: apps/desktop/src/lib/assignment-service.ts
import apiClient from "@/lib/api-client";
import type { Assignment } from "@fixly/shared-types";

export interface AssignmentsQuery {
  search?: string;
  status?: string;
  priority?: string;
  subject_id?: string;
  tags?: string[];
  is_archived?: boolean;
  is_favorite?: boolean;
  is_pinned?: boolean;
  due_date_from?: string;
  due_date_to?: string;
  sort_by?: string;
  sort_order?: string;
  page?: number;
  page_size?: number;
}

export interface PaginatedResponse {
  data: Assignment[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface AssignmentStats {
  total: number;
  completed: number;
  pending: number;
  in_progress: number;
  overdue: number;
  completion_percentage: number;
  due_today: number;
  due_this_week: number;
  avg_completion_time_hours: number | null;
}

export async function getAssignments(query: AssignmentsQuery = {}): Promise<PaginatedResponse> {
  const params = new URLSearchParams();
  Object.entries(query).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== "") {
      if (Array.isArray(value)) {
        value.forEach((v) => params.append(key, v));
      } else {
        params.set(key, String(value));
      }
    }
  });
  const response = await apiClient.get(`/api/v1/assignments?${params.toString()}`);
  return response.data;
}

export async function getAssignment(id: string): Promise<Assignment> {
  const response = await apiClient.get(`/api/v1/assignments/${id}`);
  return response.data;
}

export interface CreateAssignmentPayload {
  title: string;
  description?: string | null;
  subject_id?: string | null;
  priority?: string;
  status?: string;
  due_date?: string | null;
  estimated_study_time?: number | null;
  tags?: string[];
  notes?: string | null;
  is_pinned?: boolean;
  is_favorite?: boolean;
}

export interface UpdateAssignmentPayload {
  title?: string;
  description?: string | null;
  subject_id?: string | null;
  priority?: string;
  status?: string;
  due_date?: string | null;
  estimated_study_time?: number | null;
  tags?: string[];
  notes?: string | null;
  is_archived?: boolean;
  is_pinned?: boolean;
  is_favorite?: boolean;
}

export async function createAssignment(data: CreateAssignmentPayload): Promise<Assignment> {
  const response = await apiClient.post("/api/v1/assignments", data);
  return response.data;
}

export async function updateAssignment(id: string, data: UpdateAssignmentPayload): Promise<Assignment> {
  const response = await apiClient.put(`/api/v1/assignments/${id}`, data);
  return response.data;
}

export async function deleteAssignment(id: string): Promise<void> {
  await apiClient.delete(`/api/v1/assignments/${id}`);
}

export async function duplicateAssignment(id: string): Promise<Assignment> {
  const response = await apiClient.post(`/api/v1/assignments/${id}/duplicate`);
  return response.data;
}

export async function bulkAction(
  ids: string[],
  action: string,
  value?: string | boolean | null,
): Promise<{ affected: number; data: Assignment[] }> {
  const response = await apiClient.post("/api/v1/assignments/bulk", { ids, action, value });
  return response.data;
}

export async function getAssignmentStats(): Promise<AssignmentStats> {
  const response = await apiClient.get("/api/v1/assignments/stats");
  return response.data;
}

export async function getAssignmentAttachments(assignmentId: string): Promise<Attachment[]> {
  const response = await apiClient.get(`/api/v1/assignments/${assignmentId}/attachments`);
  return response.data;
}

export async function uploadAttachment(assignmentId: string, file: File): Promise<Attachment> {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("assignment_id", assignmentId);
  const response = await apiClient.post("/api/v1/upload", formData, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return response.data;
}

export async function deleteAttachment(attachmentId: string): Promise<void> {
  await apiClient.delete(`/api/v1/upload/${attachmentId}`);
}

interface Attachment {
  id: string;
  assignment_id: string;
  user_id: string;
  file_name: string;
  file_type: string | null;
  file_size: number | null;
  storage_path: string;
  created_at: string;
}
